<a href="https://colab.research.google.com/github/omprakash1207/sms-spam-detection/blob/main/movies_recomendation.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [259]:
import numpy as np
import pandas as pd

In [260]:
# after uploading the file into drive we have to come at google colab and then find the that file into content named file
# and then copy the address of that file and put into the bracket within single quota'' of read_csv('')
credits=pd.read_csv('/content/drive/MyDrive/Colab Notebooks/dataset/credits.csv')   # this wll  data into colab from th e drive


In [261]:
movies=pd.read_csv('/content/drive/MyDrive/Colab Notebooks/dataset/movies.csv')

In [262]:
credits.head(10)     # this will show the first 10 row of the credits

movie_id                                     title  \
0     19995                                    Avatar   
1       285  Pirates of the Caribbean: At World's End   
2    206647                                   Spectre   
3     49026                     The Dark Knight Rises   
4     49529                               John Carter   
5       559                              Spider-Man 3   
6     38757                                   Tangled   
7     99861                   Avengers: Age of Ultron   
8       767    Harry Potter and the Half-Blood Prince   
9    209112        Batman v Superman: Dawn of Justice   

                                                cast  \
0  [{"cast_id": 242, "character": "Jake Sully", "...   
1  [{"cast_id": 4, "character": "Captain Jack Spa...   
2  [{"cast_id": 1, "character": "James Bond", "cr...   
3  [{"cast_id": 2, "character": "Bruce Wayne / Ba...   
4  [{"cast_id": 5, "character": "John Carter", "c...   
5  [{"cast_id": 30, "character": "Peter Parker / ...   
6  [{"cast_id": 34, "character": "Flynn Rider (vo...   
7  [{"cast_id": 76, "character": "Tony Stark / Ir...   
8  [{"cast_id": 3, "character": "Harry Potter", "...   
9  [{"cast_id": 18, "character": "Bruce Wayne / B...   

                                                crew  
0  [{"credit_id": "52fe48009251416c750aca23", "de...  
1  [{"credit_id": "52fe4232c3a36847f800b579", "de...  
2  [{"credit_id": "54805967c3a36829b5002c41", "de...  
3  [{"credit_id": "52fe4781c3a36847f81398c3", "de...  
4  [{"credit_id": "52fe479ac3a36847f813eaa3", "de...  
5  [{"credit_id": "52fe4252c3a36847f80151a5", "de...  
6  [{"credit_id": "52fe46db9251416c91062101", "de...  
7  [{"credit_id": "55d5f7d4c3a3683e7e0016eb", "de...  
8  [{"credit_id": "52fe4273c3a36847f801fab1", "de...  
9  [{"credit_id": "553bf23692514135c8002886", "de...

In [263]:
movies.head(2)         # this will show first 2 row of the movie data

budget                                             genres  \
0  237000000  [{"id": 28, "name": "Action"}, {"id": 12, "nam...   
1  300000000  [{"id": 12, "name": "Adventure"}, {"id": 14, "...   

                                       homepage     id  \
0                   http://www.avatarmovie.com/  19995   
1  http://disney.go.com/disneypictures/pirates/    285   

                                            keywords original_language  \
0  [{"id": 1463, "name": "culture clash"}, {"id":...                en   
1  [{"id": 270, "name": "ocean"}, {"id": 726, "na...                en   

                             original_title  \
0                                    Avatar   
1  Pirates of the Caribbean: At World's End   

                                            overview  popularity  \
0  In the 22nd century, a paraplegic Marine is di...  150.437577   
1  Captain Barbossa, long believed to be dead, ha...  139.082615   

                                production_companies  \
0  [{"name": "Ingenious Film Partners", "id": 289...   
1  [{"name": "Walt Disney Pictures", "id": 2}, {"...   

                                production_countries release_date     revenue  \
0  [{"iso_3166_1": "US", "name": "United States o...   2009-12-10  2787965087   
1  [{"iso_3166_1": "US", "name": "United States o...   2007-05-19   961000000   

   runtime                                   spoken_languages    status  \
0    162.0  [{"iso_639_1": "en", "name": "English"}, {"iso...  Released   
1    169.0           [{"iso_639_1": "en", "name": "English"}]  Released   

                                          tagline  \
0                     Enter the World of Pandora.   
1  At the end of the world, the adventure begins.   

                                      title  vote_average  vote_count  
0                                    Avatar           7.2       11800  
1  Pirates of the Caribbean: At World's End           6.9        4500

In [264]:
# MERGING THE DATA OF MOVIES AND CREDITS INTO THE MOVIES  AND AFTER MERGING based on the title GIVING THE NAME IS DATA
data=movies.merge(credits,on='title')

In [265]:
# GIVE THE SHAPE OF MERGED FILE
movies.merge(credits,on='title').shape

(4809, 23)

In [266]:
# SHAPE OF THE MOVIES
movies.shape

(4803, 20)

In [267]:
# SHAPE OF THE CREDITS
credits.shape

(4803, 4)

In [268]:
# PRINTING THE FIRST TWO ROW OF MERGED FILE
data.head(2)

budget                                             genres  \
0  237000000  [{"id": 28, "name": "Action"}, {"id": 12, "nam...   
1  300000000  [{"id": 12, "name": "Adventure"}, {"id": 14, "...   

                                       homepage     id  \
0                   http://www.avatarmovie.com/  19995   
1  http://disney.go.com/disneypictures/pirates/    285   

                                            keywords original_language  \
0  [{"id": 1463, "name": "culture clash"}, {"id":...                en   
1  [{"id": 270, "name": "ocean"}, {"id": 726, "na...                en   

                             original_title  \
0                                    Avatar   
1  Pirates of the Caribbean: At World's End   

                                            overview  popularity  \
0  In the 22nd century, a paraplegic Marine is di...  150.437577   
1  Captain Barbossa, long believed to be dead, ha...  139.082615   

                                production_companies  ... runtime  \
0  [{"name": "Ingenious Film Partners", "id": 289...  ...   162.0   
1  [{"name": "Walt Disney Pictures", "id": 2}, {"...  ...   169.0   

                                    spoken_languages    status  \
0  [{"iso_639_1": "en", "name": "English"}, {"iso...  Released   
1           [{"iso_639_1": "en", "name": "English"}]  Released   

                                          tagline  \
0                     Enter the World of Pandora.   
1  At the end of the world, the adventure begins.   

                                      title vote_average vote_count movie_id  \
0                                    Avatar          7.2      11800    19995   
1  Pirates of the Caribbean: At World's End          6.9       4500      285   

                                                cast  \
0  [{"cast_id": 242, "character": "Jake Sully", "...   
1  [{"cast_id": 4, "character": "Captain Jack Spa...   

                                                crew  
0  [{"credit_id": "52fe48009251416c750aca23", "de...  
1  [{"credit_id": "52fe4232c3a36847f800b579", "de...  

[2 rows x 23 columns]

In [269]:
# finding the information of movies
movies.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4803 entries, 0 to 4802
Data columns (total 20 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   budget                4803 non-null   int64  
 1   genres                4803 non-null   object 
 2   homepage              1712 non-null   object 
 3   id                    4803 non-null   int64  
 4   keywords              4803 non-null   object 
 5   original_language     4803 non-null   object 
 6   original_title        4803 non-null   object 
 7   overview              4800 non-null   object 
 8   popularity            4803 non-null   float64
 9   production_companies  4803 non-null   object 
 10  production_countries  4803 non-null   object 
 11  release_date          4802 non-null   object 
 12  revenue               4803 non-null   int64  
 13  runtime               4801 non-null   float64
 14  spoken_languages      4803 non-null   object 
 15  status               

In [270]:
# finding the information of credits about they are  vacant or not
credits.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4803 entries, 0 to 4802
Data columns (total 4 columns):
 #   Column    Non-Null Count  Dtype 
---  ------    --------------  ----- 
 0   movie_id  4803 non-null   int64 
 1   title     4803 non-null   object
 2   cast      4803 non-null   object
 3   crew      4803 non-null   object
dtypes: int64(1), object(3)
memory usage: 150.2+ KB


In [271]:
# only these maintion column i want print , because i am working on the se only
data=data[['title','id','overview','genres','keywords','cast','crew']]

In [272]:
data.head(2)

title     id  \
0                                    Avatar  19995   
1  Pirates of the Caribbean: At World's End    285   

                                            overview  \
0  In the 22nd century, a paraplegic Marine is di...   
1  Captain Barbossa, long believed to be dead, ha...   

                                              genres  \
0  [{"id": 28, "name": "Action"}, {"id": 12, "nam...   
1  [{"id": 12, "name": "Adventure"}, {"id": 14, "...   

                                            keywords  \
0  [{"id": 1463, "name": "culture clash"}, {"id":...   
1  [{"id": 270, "name": "ocean"}, {"id": 726, "na...   

                                                cast  \
0  [{"cast_id": 242, "character": "Jake Sully", "...   
1  [{"cast_id": 4, "character": "Captain Jack Spa...   

                                                crew  
0  [{"credit_id": "52fe48009251416c750aca23", "de...  
1  [{"credit_id": "52fe4232c3a36847f800b579", "de...

In [273]:
# THIS WILL TELL ABOUT IS ANY DATA IS UNAVAILABLE OR NOT BY PRINTING FALSE IF AVILABLE AND PRINI TRUE IF NOT AVAILABLE
data.isnull()

title     id  overview  genres  keywords   cast   crew
0     False  False     False   False     False  False  False
1     False  False     False   False     False  False  False
2     False  False     False   False     False  False  False
3     False  False     False   False     False  False  False
4     False  False     False   False     False  False  False
...     ...    ...       ...     ...       ...    ...    ...
4804  False  False     False   False     False  False  False
4805  False  False     False   False     False  False  False
4806  False  False     False   False     False  False  False
4807  False  False     False   False     False  False  False
4808  False  False     False   False     False  False  False

[4809 rows x 7 columns]

In [274]:
# THIS WILL FIND AND COUNT THE HOW MANY NUMBER IS MISING IN PARTICULAR COLUMN
data.isnull().sum()

title       0
id          0
overview    3
genres      0
keywords    0
cast        0
crew        0
dtype: int64

In [275]:
#this will drop all the missing value from the available dataset
data.dropna(inplace=True)

In [276]:
# NOW CHECKING again FOR THE MISSING VALUE  and correcting it
data.isnull().sum()

title       0
id          0
overview    0
genres      0
keywords    0
cast        0
crew        0
dtype: int64

In [277]:
# NOW WE ARE LOOKING FOR ANY DUPLCATE VALUE
data.duplicated().sum()

0

In [278]:
data.head(5)

title      id  \
0                                    Avatar   19995   
1  Pirates of the Caribbean: At World's End     285   
2                                   Spectre  206647   
3                     The Dark Knight Rises   49026   
4                               John Carter   49529   

                                            overview  \
0  In the 22nd century, a paraplegic Marine is di...   
1  Captain Barbossa, long believed to be dead, ha...   
2  A cryptic message from Bond’s past sends him o...   
3  Following the death of District Attorney Harve...   
4  John Carter is a war-weary, former military ca...   

                                              genres  \
0  [{"id": 28, "name": "Action"}, {"id": 12, "nam...   
1  [{"id": 12, "name": "Adventure"}, {"id": 14, "...   
2  [{"id": 28, "name": "Action"}, {"id": 12, "nam...   
3  [{"id": 28, "name": "Action"}, {"id": 80, "nam...   
4  [{"id": 28, "name": "Action"}, {"id": 12, "nam...   

                                            keywords  \
0  [{"id": 1463, "name": "culture clash"}, {"id":...   
1  [{"id": 270, "name": "ocean"}, {"id": 726, "na...   
2  [{"id": 470, "name": "spy"}, {"id": 818, "name...   
3  [{"id": 849, "name": "dc comics"}, {"id": 853,...   
4  [{"id": 818, "name": "based on novel"}, {"id":...   

                                                cast  \
0  [{"cast_id": 242, "character": "Jake Sully", "...   
1  [{"cast_id": 4, "character": "Captain Jack Spa...   
2  [{"cast_id": 1, "character": "James Bond", "cr...   
3  [{"cast_id": 2, "character": "Bruce Wayne / Ba...   
4  [{"cast_id": 5, "character": "John Carter", "c...   

                                                crew  
0  [{"credit_id": "52fe48009251416c750aca23", "de...  
1  [{"credit_id": "52fe4232c3a36847f800b579", "de...  
2  [{"credit_id": "54805967c3a36829b5002c41", "de...  
3  [{"credit_id": "52fe4781c3a36847f81398c3", "de...  
4  [{"credit_id": "52fe479ac3a36847f813eaa3", "de...

In [279]:
#ILOC FUNCTION IS USED TO LOCATE THE PARTICULAR ROW AND COLUMN(LIST) by giving their particular number of row and column
data.iloc[0,3]    # column no of genres ==3

'[{"id": 28, "name": "Action"}, {"id": 12, "name": "Adventure"}, {"id": 14, "name": "Fantasy"}, {"id": 878, "name": "Science Fiction"}]'

In [280]:
# BOTH WILL REMAIN SAME
data.iloc[0].genres

'[{"id": 28, "name": "Action"}, {"id": 12, "name": "Adventure"}, {"id": 14, "name": "Fantasy"}, {"id": 878, "name": "Science Fiction"}]'

BUT WE WANT ONLY 'ACTION',  'ADVENTURE','FANTASY','SCIENCE FICTION ' ONLY

means they have the

In [281]:
#   converting the

In [282]:
data.iloc[0,4]

'[{"id": 1463, "name": "culture clash"}, {"id": 2964, "name": "future"}, {"id": 3386, "name": "space war"}, {"id": 3388, "name": "space colony"}, {"id": 3679, "name": "society"}, {"id": 3801, "name": "space travel"}, {"id": 9685, "name": "futuristic"}, {"id": 9840, "name": "romance"}, {"id": 9882, "name": "space"}, {"id": 9951, "name": "alien"}, {"id": 10148, "name": "tribe"}, {"id": 10158, "name": "alien planet"}, {"id": 10987, "name": "cgi"}, {"id": 11399, "name": "marine"}, {"id": 13065, "name": "soldier"}, {"id": 14643, "name": "battle"}, {"id": 14720, "name": "love affair"}, {"id": 165431, "name": "anti war"}, {"id": 193554, "name": "power relations"}, {"id": 206690, "name": "mind and soul"}, {"id": 209714, "name": "3d"}]'

In [283]:
import ast

In [284]:
# CREATING A HELPER FUNCTION WHICH MAINLLY  STORE THE SOME VALUE FROM DICTIONARY AND PUT INTO THE LIST

def convert(object):      # DEFINING A CONVERTER WHICH WILL MANLY CONVERT  A DICTONARY INTO  THE LIST  AS WE CAN RUN THE LOOP ON LIST
  L=[]    # CREATING  A LIST
  count=0
  for i in ast.literal_eval(object):
    if count!=3:
      L.append(i['name'])        # THIS  WILL MAINLY ADD ALL THE NAME RELATED VALUE IN THE LIST L
    else:
      break
    return L         # THIS WILL RETURN THE LIST L

In [285]:
data['genres']=data['genres'].apply(convert)

In [286]:
data['keywords']=data['keywords'].apply(convert)

In [287]:
data.head(3)

title      id  \
0                                    Avatar   19995   
1  Pirates of the Caribbean: At World's End     285   
2                                   Spectre  206647   

                                            overview       genres  \
0  In the 22nd century, a paraplegic Marine is di...     [Action]   
1  Captain Barbossa, long believed to be dead, ha...  [Adventure]   
2  A cryptic message from Bond’s past sends him o...     [Action]   

          keywords                                               cast  \
0  [culture clash]  [{"cast_id": 242, "character": "Jake Sully", "...   
1          [ocean]  [{"cast_id": 4, "character": "Captain Jack Spa...   
2            [spy]  [{"cast_id": 1, "character": "James Bond", "cr...   

                                                crew  
0  [{"credit_id": "52fe48009251416c750aca23", "de...  
1  [{"credit_id": "52fe4232c3a36847f800b579", "de...  
2  [{"credit_id": "54805967c3a36829b5002c41", "de...

In [288]:
data['cast'][0]

'[{"cast_id": 242, "character": "Jake Sully", "credit_id": "5602a8a7c3a3685532001c9a", "gender": 2, "id": 65731, "name": "Sam Worthington", "order": 0}, {"cast_id": 3, "character": "Neytiri", "credit_id": "52fe48009251416c750ac9cb", "gender": 1, "id": 8691, "name": "Zoe Saldana", "order": 1}, {"cast_id": 25, "character": "Dr. Grace Augustine", "credit_id": "52fe48009251416c750aca39", "gender": 1, "id": 10205, "name": "Sigourney Weaver", "order": 2}, {"cast_id": 4, "character": "Col. Quaritch", "credit_id": "52fe48009251416c750ac9cf", "gender": 2, "id": 32747, "name": "Stephen Lang", "order": 3}, {"cast_id": 5, "character": "Trudy Chacon", "credit_id": "52fe48009251416c750ac9d3", "gender": 1, "id": 17647, "name": "Michelle Rodriguez", "order": 4}, {"cast_id": 8, "character": "Selfridge", "credit_id": "52fe48009251416c750ac9e1", "gender": 2, "id": 1771, "name": "Giovanni Ribisi", "order": 5}, {"cast_id": 7, "character": "Norm Spellman", "credit_id": "52fe48009251416c750ac9dd", "gender": 

In [289]:
data.cast.head(10)     # JUST LOOKING THE CAST COLUMN

0    [{"cast_id": 242, "character": "Jake Sully", "...
1    [{"cast_id": 4, "character": "Captain Jack Spa...
2    [{"cast_id": 1, "character": "James Bond", "cr...
3    [{"cast_id": 2, "character": "Bruce Wayne / Ba...
4    [{"cast_id": 5, "character": "John Carter", "c...
5    [{"cast_id": 30, "character": "Peter Parker / ...
6    [{"cast_id": 34, "character": "Flynn Rider (vo...
7    [{"cast_id": 76, "character": "Tony Stark / Ir...
8    [{"cast_id": 3, "character": "Harry Potter", "...
9    [{"cast_id": 18, "character": "Bruce Wayne / B...
Name: cast, dtype: object

since cast having lots information and data into this  so we have to seperately define a function which take maximum input data of 3 in one row of cast

In [290]:
#

def convert2(obj):
  m=[]                   # CREATING A EMPTY LIST
  counter = 0
  for i in ast.literal_eval(obj):
    if counter !=3:     # only upto three value will be appended
      m.append(i['name'])
      counter +=1    # increasing the counter
    else :
      break
    return m

In [291]:
data['cast']=data['cast'].apply(convert2)



In [292]:
data.head(3)

title      id  \
0                                    Avatar   19995   
1  Pirates of the Caribbean: At World's End     285   
2                                   Spectre  206647   

                                            overview       genres  \
0  In the 22nd century, a paraplegic Marine is di...     [Action]   
1  Captain Barbossa, long believed to be dead, ha...  [Adventure]   
2  A cryptic message from Bond’s past sends him o...     [Action]   

          keywords               cast  \
0  [culture clash]  [Sam Worthington]   
1          [ocean]      [Johnny Depp]   
2            [spy]     [Daniel Craig]   

                                                crew  
0  [{"credit_id": "52fe48009251416c750aca23", "de...  
1  [{"credit_id": "52fe4232c3a36847f800b579", "de...  
2  [{"credit_id": "54805967c3a36829b5002c41", "de...

In [293]:
data['crew'][0]

'[{"credit_id": "52fe48009251416c750aca23", "department": "Editing", "gender": 0, "id": 1721, "job": "Editor", "name": "Stephen E. Rivkin"}, {"credit_id": "539c47ecc3a36810e3001f87", "department": "Art", "gender": 2, "id": 496, "job": "Production Design", "name": "Rick Carter"}, {"credit_id": "54491c89c3a3680fb4001cf7", "department": "Sound", "gender": 0, "id": 900, "job": "Sound Designer", "name": "Christopher Boyes"}, {"credit_id": "54491cb70e0a267480001bd0", "department": "Sound", "gender": 0, "id": 900, "job": "Supervising Sound Editor", "name": "Christopher Boyes"}, {"credit_id": "539c4a4cc3a36810c9002101", "department": "Production", "gender": 1, "id": 1262, "job": "Casting", "name": "Mali Finn"}, {"credit_id": "5544ee3b925141499f0008fc", "department": "Sound", "gender": 2, "id": 1729, "job": "Original Music Composer", "name": "James Horner"}, {"credit_id": "52fe48009251416c750ac9c3", "department": "Directing", "gender": 2, "id": 2710, "job": "Director", "name": "James Cameron"},

In [294]:
# CREATING A FUNCTION WHOSE NAME IS FETCH_DIRECTOR

def fetch_director(obj):
  n=[]
  for i in ast.literal_eval(obj):
    if i ['job']=='Director':
      n.append(i['name'])
      break
  return n

In [295]:
data['crew']=data['crew'].apply(fetch_director)     # RECALLING FETCH_DIRECTOR  FUNCTION

In [296]:
data.head(3)

title      id  \
0                                    Avatar   19995   
1  Pirates of the Caribbean: At World's End     285   
2                                   Spectre  206647   

                                            overview       genres  \
0  In the 22nd century, a paraplegic Marine is di...     [Action]   
1  Captain Barbossa, long believed to be dead, ha...  [Adventure]   
2  A cryptic message from Bond’s past sends him o...     [Action]   

          keywords               cast              crew  
0  [culture clash]  [Sam Worthington]   [James Cameron]  
1          [ocean]      [Johnny Depp]  [Gore Verbinski]  
2            [spy]     [Daniel Craig]      [Sam Mendes]

In [297]:
#   COLUMN OF OVERVIEW IS A STRING since this is string so we have to convert into the list so that we can concatenate

data['overview'][0]

'In the 22nd century, a paraplegic Marine is dispatched to the moon Pandora on a unique mission, but becomes torn between following orders and protecting an alien civilization.'

In [298]:
#  WE APPLYING A LEMBDA FUNCTION to convert string into a list

data['overview']=data['overview'].apply(lambda x:x.split())    #

In [299]:
data.head()

title      id  \
0                                    Avatar   19995   
1  Pirates of the Caribbean: At World's End     285   
2                                   Spectre  206647   
3                     The Dark Knight Rises   49026   
4                               John Carter   49529   

                                            overview       genres  \
0  [In, the, 22nd, century,, a, paraplegic, Marin...     [Action]   
1  [Captain, Barbossa,, long, believed, to, be, d...  [Adventure]   
2  [A, cryptic, message, from, Bond’s, past, send...     [Action]   
3  [Following, the, death, of, District, Attorney...     [Action]   
4  [John, Carter, is, a, war-weary,, former, mili...     [Action]   

           keywords               cast                 crew  
0   [culture clash]  [Sam Worthington]      [James Cameron]  
1           [ocean]      [Johnny Depp]     [Gore Verbinski]  
2             [spy]     [Daniel Craig]         [Sam Mendes]  
3       [dc comics]   [Christian Bale]  [Christopher Nolan]  
4  [based on novel]    [Taylor Kitsch]     [Andrew Stanton]

In [300]:
data['genres'].info()

<class 'pandas.core.series.Series'>
Index: 4806 entries, 0 to 4808
Series name: genres
Non-Null Count  Dtype 
--------------  ----- 
4778 non-null   object
dtypes: object(1)
memory usage: 204.1+ KB


In [301]:
# we have to apply a conversion due to make a unique word in each column  ,
#jenerally we remove the space between names SO THAT EACH NAME WILL MAKAE UNIQUE NAME


In [302]:
data['genres'].info

<bound method Series.info of 0            [Action]
1         [Adventure]
2            [Action]
3            [Action]
4            [Action]
            ...      
4804         [Action]
4805         [Comedy]
4806         [Comedy]
4807             None
4808    [Documentary]
Name: genres, Length: 4806, dtype: object>

In [303]:
data.dropna(inplace=True)

In [304]:
# THIS IS THE METHOD OF REMOVING THE EXTRA SPACE FROM THE WORD


data['genres']   =data['genres'].apply(lambda x : [i.replace(" ","") for i in x ])
data['keywords'] =data['keywords'].apply(lambda x : [i.replace(" ","") for i in x ])
data['cast']     =data['cast'].apply(lambda x : [i.replace(" ","") for i in x ])
data['crew']     =data['crew'].apply(lambda x : [i.replace(" ","") for i in x ])

In [305]:
data.head(5)

title      id  \
0                                    Avatar   19995   
1  Pirates of the Caribbean: At World's End     285   
2                                   Spectre  206647   
3                     The Dark Knight Rises   49026   
4                               John Carter   49529   

                                            overview       genres  \
0  [In, the, 22nd, century,, a, paraplegic, Marin...     [Action]   
1  [Captain, Barbossa,, long, believed, to, be, d...  [Adventure]   
2  [A, cryptic, message, from, Bond’s, past, send...     [Action]   
3  [Following, the, death, of, District, Attorney...     [Action]   
4  [John, Carter, is, a, war-weary,, former, mili...     [Action]   

         keywords              cast                crew  
0  [cultureclash]  [SamWorthington]      [JamesCameron]  
1         [ocean]      [JohnnyDepp]     [GoreVerbinski]  
2           [spy]     [DanielCraig]         [SamMendes]  
3      [dccomics]   [ChristianBale]  [ChristopherNolan]  
4  [basedonnovel]    [TaylorKitsch]     [AndrewStanton]

In [306]:
# MAKING A NEW COLUMN NAME 'TAG' AND ADDING ALL THE COLUMN AS MENTIONED

In [307]:
#data  =  data['overview'] + data['genres'] + data['keywords	'] + data['cast'] + data['crew']             # this is not working


In [308]:
data['tags'] = data['overview'] + data['genres']  + data['keywords'] + data['cast']  + data['crew']

In [309]:
# ANOTHER METHOD

 # Concatenate  all different list into a single list
#data['tags'] = data[[ 'overview','genres', 'keywords', 'cast', 'crew']].values.tolist()

In [310]:
data.head()

title      id  \
0                                    Avatar   19995   
1  Pirates of the Caribbean: At World's End     285   
2                                   Spectre  206647   
3                     The Dark Knight Rises   49026   
4                               John Carter   49529   

                                            overview       genres  \
0  [In, the, 22nd, century,, a, paraplegic, Marin...     [Action]   
1  [Captain, Barbossa,, long, believed, to, be, d...  [Adventure]   
2  [A, cryptic, message, from, Bond’s, past, send...     [Action]   
3  [Following, the, death, of, District, Attorney...     [Action]   
4  [John, Carter, is, a, war-weary,, former, mili...     [Action]   

         keywords              cast                crew  \
0  [cultureclash]  [SamWorthington]      [JamesCameron]   
1         [ocean]      [JohnnyDepp]     [GoreVerbinski]   
2           [spy]     [DanielCraig]         [SamMendes]   
3      [dccomics]   [ChristianBale]  [ChristopherNolan]   
4  [basedonnovel]    [TaylorKitsch]     [AndrewStanton]   

                                                tags  
0  [In, the, 22nd, century,, a, paraplegic, Marin...  
1  [Captain, Barbossa,, long, believed, to, be, d...  
2  [A, cryptic, message, from, Bond’s, past, send...  
3  [Following, the, death, of, District, Attorney...  
4  [John, Carter, is, a, war-weary,, former, mili...

In [311]:
# creating a  newdata which contain "id","title","tags"
newdata = data[['id','title', 'tags' ]]

In [312]:
newdata.head()

id                                     title  \
0   19995                                    Avatar   
1     285  Pirates of the Caribbean: At World's End   
2  206647                                   Spectre   
3   49026                     The Dark Knight Rises   
4   49529                               John Carter   

                                                tags  
0  [In, the, 22nd, century,, a, paraplegic, Marin...  
1  [Captain, Barbossa,, long, believed, to, be, d...  
2  [A, cryptic, message, from, Bond’s, past, send...  
3  [Following, the, death, of, District, Attorney...  
4  [John, Carter, is, a, war-weary,, former, mili...

In [313]:
 # we have to join the list at each space by converting list into a string
newdata['tags']=newdata['tags'].apply(lambda x:" ".join(x))

<ipython-input-313-6b82a7e81761>:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  newdata['tags']=newdata['tags'].apply(lambda x:" ".join(x))


In [314]:
newdata['tags'][0]

'In the 22nd century, a paraplegic Marine is dispatched to the moon Pandora on a unique mission, but becomes torn between following orders and protecting an alien civilization. Action cultureclash SamWorthington JamesCameron'

In [315]:
newdata.head()

id                                     title  \
0   19995                                    Avatar   
1     285  Pirates of the Caribbean: At World's End   
2  206647                                   Spectre   
3   49026                     The Dark Knight Rises   
4   49529                               John Carter   

                                                tags  
0  In the 22nd century, a paraplegic Marine is di...  
1  Captain Barbossa, long believed to be dead, ha...  
2  A cryptic message from Bond’s past sends him o...  
3  Following the death of District Attorney Harve...  
4  John Carter is a war-weary, former military ca...

NOW WE HAVE MAKE A SITE AT WHICH WHEN WE GIVE ANY MOVIE NAME THEN OUR PROJECT WILL RECOMMEND 5 SIMILLAR NAME OF MOVIE OR MOST SIMILAR NAME  

THERE IS MANY METHODE TO DO THIS
LIKE CAMPARE ALL THE MOVIE NAME WITH STORED MOVIE NNAME IN THE OUT DATASET

SO TO AVOID COMPLICATION WE JUST GO FOR VECTORIZATION SO THAT IT WILL GIVE A PARTICULAR MOVIE CORRESPONDING COORDINATE AND EASY TO LOCATE A MOVIE


NOW OUT OF LARGE WORDS WE WILL JUST FIND MOST COMMON FREQUANTLY COMMING WORDS
AND GIVE IT TO AS VECTOR Y AXIS AND MOVIE IN X AXIS   

DEURING  THE VECTORIZATION WE DID NOT CONSIDER STOP WORD WHICH I S USE D TO MAKE A  SENTENCE

In [316]:
# TO DELETE THE STOP WORD  WE USE LIBRARY LIKE SKLEARN   countvector class will give us which one is the most frequent
# word comming in the list of arraged string
from sklearn.feature_extraction.text import CountVectorizer    # here we are removing stop word of english language
cv=CountVectorizer(max_features = 5000, stop_words='english')


In [317]:
cv.fit_transform(newdata['tags']).toarray().shape

(4384, 5000)

In [318]:
cv.fit_transform(newdata['tags']).toarray()    # converting vector into array (numpy array )

array([[0, 0, 0, ..., 0, 0, 0],
       [0, 0, 0, ..., 0, 0, 0],
       [0, 0, 0, ..., 0, 0, 0],
       ...,
       [0, 0, 0, ..., 0, 0, 0],
       [0, 0, 0, ..., 0, 0, 0],
       [0, 0, 0, ..., 0, 0, 0]])

In [319]:
vectors=cv.fit_transform(newdata['tags']).toarray()

In [320]:
vectors[0]

array([0, 0, 0, ..., 0, 0, 0])

In [321]:
cv.get_feature_names_out()

array(['000', '007', '10', ..., 'zombies', 'zone', 'zoo'], dtype=object)

In [322]:
len(cv.get_feature_names_out())

5000

In [323]:
# stemming applying  means ['loved','loving ','love ']  ----> converted into  ['love','love','love']
# to reduce the repeating word from

In [324]:
# we have to have to import some library to do this

In [325]:
newdata['tags'][0]

'In the 22nd century, a paraplegic Marine is dispatched to the moon Pandora on a unique mission, but becomes torn between following orders and protecting an alien civilization. Action cultureclash SamWorthington JamesCameron'

In [326]:
import nltk
from nltk.stem.porter import PorterStemmer
ps = PorterStemmer()
def stem(text):
    y = []

    for i in text.split():
        y.append(ps.stem(i))
    return " ".join(y)

In [327]:
newdata['tags'] = newdata['tags'].apply(stem)

<ipython-input-327-b46c1f9cb2bb>:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  newdata['tags'] = newdata['tags'].apply(stem)


In [328]:
#newdata['stemmed_tags']

In [329]:
cv.get_feature_names_out()

array(['000', '007', '10', ..., 'zombies', 'zone', 'zoo'], dtype=object)

In [330]:
cv.get_feature_names_out().shape

(5000,)

we have total 4806 movies so we have 4806 vector also so
we will find the distance between two vector(nuclidian distance )
as angle
if angle is more ration or simillarity is low  ================
cosine similarity in sikite learn .

In [331]:
import sklearn
from sklearn.metrics.pairwise import cosine_similarity

In [332]:
cosine_similarity(vectors).shape

(4384, 4384)

In [333]:
similarity=cosine_similarity(vectors)

In [334]:

similarity [0]

array([1.        , 0.        , 0.05163978, ..., 0.04364358, 0.        ,
       0.        ])

In [335]:
# if we have provide any movie then we have to find a movie endex from the vector which we hve to provide the name

In [336]:
newdata

id                                     title  \
0      19995                                    Avatar   
1        285  Pirates of the Caribbean: At World's End   
2     206647                                   Spectre   
3      49026                     The Dark Knight Rises   
4      49529                               John Carter   
...      ...                                       ...   
4801  124606                                      Bang   
4802   14337                                    Primer   
4804    9367                               El Mariachi   
4806  231617                 Signed, Sealed, Delivered   
4808   25975                         My Date with Drew   

                                                   tags  
0     in the 22nd century, a parapleg marin is dispa...  
1     captain barbossa, long believ to be dead, ha c...  
2     a cryptic messag from bond’ past send him on a...  
3     follow the death of district attorney harvey d...  
4     john carter is a war-weary, former militari ca...  
...                                                 ...  
4801  a young woman in l.a. is have a bad day: she' ...  
4802  friends/fledgl entrepreneur invent a devic in ...  
4804  el mariachi just want to play hi guitar and ca...  
4806  "signed, sealed, delivered" introduc a dedic q...  
4808  ever sinc the second grade when he first saw h...  

[4384 rows x 3 columns]

In [337]:
newdata[newdata['title']== 'Avatar'].index[0]   # this will give index of movie

0

In [338]:
# here we are comparing the movie with start 0 to end 5000 but due to sorting index position will be changes so we have to use differen t function
sorted(similarity[0],  reverse = True)

[0.9999999999999998,
 0.2608745973749755,
 0.21081851067789195,
 0.19462473604038075,
 0.17978662999019787,
 0.1732050807568877,
 0.16151457061744964,
 0.15491933384829668,
 0.15491933384829668,
 0.14907119849998599,
 0.14907119849998599,
 0.14638501094227996,
 0.14462030521243743,
 0.1438389904456152,
 0.13912166872805048,
 0.13912166872805048,
 0.13801311186847084,
 0.13801311186847084,
 0.13801311186847084,
 0.13801311186847084,
 0.1369306393762915,
 0.1348399724926484,
 0.1333333333333333,
 0.1333333333333333,
 0.1333333333333333,
 0.1333333333333333,
 0.1333333333333333,
 0.1333333333333333,
 0.13223592098145726,
 0.1309307341415954,
 0.12909944487358055,
 0.12909944487358055,
 0.12909944487358055,
 0.12909944487358055,
 0.12734290799340264,
 0.12524485821702988,
 0.12403473458920844,
 0.1224744871391589,
 0.1224744871391589,
 0.12171612389003691,
 0.12171612389003691,
 0.11846977555181845,
 0.11846977555181845,
 0.11677484162422844,
 0.11619857875482535,
 0.11547005383792514,
 0.

In [339]:
enumerate(similarity[0])    # we will get enumerate object

In [340]:
list(enumerate(similarity[0]))   # using the enumerate fn we are putting their index along with them also
# in ans showing  0 th movie ka  other movie se distance along with their index

[(0, 0.9999999999999998),
 (1, 0.0),
 (2, 0.051639777949432225),
 (3, 0.07093269021319086),
 (4, 0.047946330148538406),
 (5, 0.046373889576016826),
 (6, 0.03849001794597505),
 (7, 0.13912166872805048),
 (8, 0.0),
 (9, 0.04364357804719847),
 (10, 0.0),
 (11, 0.04494665749754947),
 (12, 0.0),
 (13, 0.03984095364447979),
 (14, 0.05270462766947299),
 (15, 0.0),
 (16, 0.0),
 (17, 0.0),
 (18, 0.03546634510659543),
 (19, 0.04134491152973615),
 (20, 0.031544014893825585),
 (21, 0.051639777949432225),
 (22, 0.0),
 (23, 0.0),
 (24, 0.0),
 (25, 0.032791291789197645),
 (26, 0.0563436169819011),
 (27, 0.07532435771547094),
 (28, 0.06900655593423542),
 (29, 0.07874992309581577),
 (30, 0.044280744277004756),
 (31, 0.12909944487358055),
 (32, 0.0),
 (33, 0.0),
 (34, 0.0),
 (35, 0.0),
 (36, 0.09274777915203365),
 (37, 0.0),
 (38, 0.033058980245364314),
 (39, 0.0),
 (40, 0.0),
 (41, 0.0),
 (42, 0.0),
 (43, 0.09274777915203365),
 (44, 0.0816496580927726),
 (45, 0.11065666703449761),
 (46, 0.0577350269189

In [341]:
sorted(list(enumerate(similarity[0])),reverse=True,key=lambda x:x[1])   # this will show  that we have to compared on the basis of 2nd index movie

[(0, 0.9999999999999998),
 (3399, 0.2608745973749755),
 (149, 0.21081851067789195),
 (524, 0.19462473604038075),
 (309, 0.17978662999019787),
 (563, 0.1732050807568877),
 (1584, 0.16151457061744964),
 (960, 0.15491933384829668),
 (3412, 0.15491933384829668),
 (565, 0.14907119849998599),
 (961, 0.14907119849998599),
 (1318, 0.14638501094227996),
 (148, 0.14462030521243743),
 (1144, 0.1438389904456152),
 (7, 0.13912166872805048),
 (73, 0.13912166872805048),
 (816, 0.13801311186847084),
 (1504, 0.13801311186847084),
 (2493, 0.13801311186847084),
 (3795, 0.13801311186847084),
 (3250, 0.1369306393762915),
 (1769, 0.1348399724926484),
 (68, 0.1333333333333333),
 (132, 0.1333333333333333),
 (2077, 0.1333333333333333),
 (2672, 0.1333333333333333),
 (2917, 0.1333333333333333),
 (3918, 0.1333333333333333),
 (125, 0.13223592098145726),
 (1174, 0.1309307341415954),
 (31, 0.12909944487358055),
 (146, 0.12909944487358055),
 (273, 0.12909944487358055),
 (490, 0.12909944487358055),
 (147, 0.1273429079

In [342]:
sorted(list(enumerate(similarity[0])),reverse=True,key=lambda x:x[1])[1:5]   # give first five movie most similar

[(3399, 0.2608745973749755),
 (149, 0.21081851067789195),
 (524, 0.19462473604038075),
 (309, 0.17978662999019787)]

In [343]:
# creating a function such a way that when a movie name given then function will  recommend 5 related movie

def recommend(movies):
  movie_index=newdata[newdata['title']== 'Avatar'].index[0]
  distances = similarity[movie_index]
  movie_list = sorted(list(enumerate(similarity[0])),reverse=True,key=lambda x:x[1])[1:5]
  for i in movie_list:
    print(i[0])



In [344]:
recommend('Alexander  ')   # this will give the index of the  movie recommended

3399
149
524
309


In [345]:
#  but as we need the name of the movie  so we go for the iloc to find the title for corresponding index
newdata.iloc[ 110 ].title

'Alexander'

In [346]:
 #  so we are just modify the upper function and insert the iloc function to get title
def recommended(movies):
  movie_index=newdata[newdata['title']== 'Avatar'].index[0]
  distances = similarity[movie_index]
  movie_list = sorted(list(enumerate(similarity[0])),reverse=True,key=lambda x:x[1])[1:5]
  for i in movie_list:
    print( newdata.iloc[i[0]].title )


In [347]:
recommended('Avatar')

Apollo 18
Beowulf
Tears of the Sun
The Adventures of Pluto Nash
